mono/mini/mini-posix.c

```diff
@@ -521,120 +504,58 @@ mono_runtime_cleanup_handlers (void)
+	guint64 us = 1000000 / mono_profiler_get_sampling_rate () - 1;
+
+	while (InterlockedRead (&sampling_thread_running)) {
+		FOREACH_THREAD_SAFE (info) {
+			if (mono_thread_info_get_tid (info) == me)
+				continue;
+
+			mono_threads_pthread_kill (info, SIGPROF);
+		} FOREACH_THREAD_SAFE_END
+
+		mono_thread_info_usleep (us);
 	}
-	return 1;
+
+	return NULL;
 }
-#endif
 
 void
 mono_runtime_shutdown_stat_profiler (void)
 {
-#ifdef HAVE_LINUX_RTC_H
-	if (rtc_fd >= 0)
-		enable_rtc_timer (FALSE);
-#endif
-}
-
-#ifdef ITIMER_PROF
-static int
-get_itimer_mode (void)
-{
-	switch (mono_profiler_get_sampling_mode ()) {
-	case MONO_PROFILER_STAT_MODE_PROCESS: return ITIMER_PROF;
-	case MONO_PROFILER_STAT_MODE_REAL: return ITIMER_REAL;
-	}
-	g_assert_not_reached ();
-	return 0;
-}
+#ifdef SIGPROF
+	InterlockedWrite (&sampling_thread_running, 0);
+	pthread_join (sampling_thread, NULL);
 
-static int
-get_itimer_signal (void)
-{
-	switch (mono_profiler_get_sampling_mode ()) {
-	case MONO_PROFILER_STAT_MODE_PROCESS: return SIGPROF;
-	case MONO_PROFILER_STAT_MODE_REAL: return SIGALRM;
-	}
-	g_assert_not_reached ();
-	return 0;
-}
+	/*
+	 * FIXME: We can't safely remove the signal handler because we have no
+	 * guarantee that all pending signals have been delivered at this point.
+	 */
+	//remove_signal_handler (SIGPROF);
 #endif
+}
 
 void
 mono_runtime_setup_stat_profiler (void)
 {
-#ifdef ITIMER_PROF
-	struct itimerval itval;
-	static int inited = 0;
-#ifdef HAVE_LINUX_RTC_H
-	const char *rtc_freq;
-	if (!inited && (rtc_freq = g_getenv ("MONO_RTC"))) {
-		int freq = 0;
-		inited = 1;
-		if (*rtc_freq)
-			freq = atoi (rtc_freq);
-		if (!freq)
-			freq = 1024;
-		rtc_fd = open ("/dev/rtc", O_RDONLY);
-		if (rtc_fd == -1) {
-			perror ("open /dev/rtc");
-			return;
-		}
-		profiling_signal_in_use = SIGPROF;
-		add_signal_handler (profiling_signal_in_use, sigprof_signal_handler, SA_RESTART);
-		if (ioctl (rtc_fd, RTC_IRQP_SET, freq) == -1) {
-			perror ("set rtc freq");
-			return;
-		}
-		if (ioctl (rtc_fd, RTC_PIE_ON, 0) == -1) {
-			perror ("start rtc");
-			return;
-		}
-		if (fcntl (rtc_fd, F_SETSIG, SIGPROF) == -1) {
-			perror ("setsig");
-			return;
-		}
-		if (fcntl (rtc_fd, F_SETOWN, getpid ()) == -1) {
-			perror ("setown");
-			return;
-		}
-		enable_rtc_timer (TRUE);
-		return;
-	}
-	if (rtc_fd >= 0)
-		return;
-#endif
+#ifdef SIGPROF
+	add_signal_handler (SIGPROF, sigprof_signal_handler, SA_RESTART);
 
-	itval.it_interval.tv_usec = (1000000 / mono_profiler_get_sampling_rate ()) - 1;
-	itval.it_interval.tv_sec = 0;
-	itval.it_value = itval.it_interval;
-	if (inited)
-		return;
-	inited = 1;
-	profiling_signal_in_use = get_itimer_signal ();
-	add_signal_handler (profiling_signal_in_use, sigprof_signal_handler, SA_RESTART);
-	setitimer (get_itimer_mode (), &itval, NULL);
+	InterlockedWrite (&sampling_thread_running, 1);
+	mono_native_thread_create (&sampling_thread, sampling_thread_func, NULL);
```

Comment:
I did it this way based on precedent: The (single) SGen thread pool thread does exactly what I'm doing here (create the thread via `mono_native_thread_create` and call `mono_thread_info_register_small_id` to get access to the lock-free data structures). It made sense to me since there's probably no reason we should want the sampler thread in the thread list (especially if we don't use `mono_thread_info_usleep` after all, as discussed above). That said, I don't actually feel strongly about it either way. We can certainly just call `mono_threads_attach_tools_thread` for simplicity's sake.

But to be clear, there's not actually anything wrong with the way the thread is *created*, right? Just the way it's being attached? Your original comment talked about suspendability, but I don't think it would be suspendable with either approach.